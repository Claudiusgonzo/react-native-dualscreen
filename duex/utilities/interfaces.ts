import React, { ReactElement } from 'react';
import { IHeader } from '../Shared/screenStore/headerStore/header.interface';

export interface IScreenComponent {
  key: string;
  duexElement: ReactElement;
  header?: IHeader;
  screen?: screenType;
}

export interface IScreenState {
  screen: Array<IScreenComponent>;
}


export interface IEmptyAction {
  type: string;
  payload: {
  };
}

export enum screenType {
  SINGLE = 'SINGLE',
  DUAL = 'DUAL'
}
