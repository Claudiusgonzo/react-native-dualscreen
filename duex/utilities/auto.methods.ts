import utility from './utility.methods';
import dualScreen from '../dualScreen/dualScreenStore/dualScreen.methods';
import singleScreen from '../singleScreen/singleScreenStore/singleScreen.methods';
import { ReactElement } from 'react';
import { IHeader } from '../Shared/screenStore/headerStore/header.interface';


/**
 * Automatically pushes element to the top of the stack based on screen size.
 */
const Add = (key: string, element: ReactElement, header?: IHeader, isMergeSingle = false, isMergeDual = false) => {
  if (utility.isDualScreen()) {
    dualScreen.Add(key, element, header, isMergeDual);
  } else {
    singleScreen.Add(key, element, header, isMergeSingle);
  }
};

/**
 * Automatically pushes element to the top of stack or if the key is already in the stack,
    move that key to the top of the stack based on screen size
 */
const AddOrMoveToFront = (key: string, element: ReactElement, header?: IHeader, isMergeSingle = false, isMergeDual = false,) => {
  if (utility.isDualScreen()) {
    dualScreen.AddOrMoveToFront(key, element, header, isMergeDual);
  } else {
    singleScreen.AddOrMoveToFront(key, element, header, isMergeSingle);
  }
};

/**
 * Automatically pushes element to the top of stack or if the key is already in the  dual screen stack,
    move that key to the top of the dual screen stack based on screen size
 */
const AddOrMoveToFrontDual = (key: string, element: ReactElement, header?: IHeader, isMergeSingle = false, isMergeDual = false,) => {
  if (utility.isDualScreen()) {
    dualScreen.AddOrMoveToFront(key, element, header, isMergeDual);
  } else {
    singleScreen.Add(key, element, header, isMergeSingle);
  }
};

/**
 * Automatically pushes element to the top of stack or if the key is already in the single screen stack,
    move that key to the top of the single screen stack based on screen size
 */
const AddOrMoveToFrontSingle = (key: string, element: ReactElement, header?: IHeader, isMergeSingle = false, isMergeDual = false,) => {
  if (utility.isDualScreen()) {
    dualScreen.Add(key, element, header, isMergeDual);
  } else {
    singleScreen.AddOrMoveToFront(key, element, header, isMergeSingle);
  }
};

/**
 * Automatically Removes all elements of the stack and returns the base element of the screen stack based on screen size
 */
const BackToHome = () => {
  if (utility.isDualScreen()) {
    dualScreen.BackToHome();
  } else {
    singleScreen.BackToHome();
  }
};


/**
 * Automatically go back one element in the stack based on screen size
 */
const GoBack = () => {
  if (utility.isDualScreen()) {
    dualScreen.GoBack();
  } else {
    singleScreen.GoBack();
  }
};

/**
 * Automatically replace element in the stack based on screen size
 */
const ReplaceScreen = (key: string, duexElement: React.ReactElement) => {
  if (utility.isDualScreen()) {
    dualScreen.ReplaceScreen(key, duexElement);
  } else {
    singleScreen.ReplaceScreen(key, duexElement);
  }
};


/**
 * Automatically replace header in the stack based on screen size
 */
const ReplaceHeader = (key: string, header: IHeader) => {
  if (utility.isDualScreen()) {
    dualScreen.ReplaceHeader(key, header);
  } else {
    singleScreen.ReplaceHeader(key, header);
  }
};



const _autoScreenFunctions = {
  Add,
  AddOrMoveToFront,
  AddOrMoveToFrontDual,
  AddOrMoveToFrontSingle,
  BackToHome,
  GoBack,
  ReplaceScreen,
  ReplaceHeader
};

type autoScreenFunctions = typeof _autoScreenFunctions;

const autoScreen: autoScreenFunctions = _autoScreenFunctions;

export default autoScreen;
