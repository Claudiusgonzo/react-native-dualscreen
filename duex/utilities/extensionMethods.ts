

interface Array<T> {
    /**
  * Returns the last element from an array.
  */
    peek(): T;
}

Array.prototype.peek = function () {
    if (this.length > 0) {
        return this[this.length - 1];
    }
    // empty array...
    return undefined;    // or another default value...
};

