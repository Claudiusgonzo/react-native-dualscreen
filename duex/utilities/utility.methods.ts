import { useRef, useEffect } from 'react';
import { store } from '../appStore';

const isDualScreen = (): boolean => {
  return store.getState().utilityStoreReducer.isDualScreen
};

export const usePrevious = <T extends {}>(value: T): T | undefined => {
  const ref = useRef<T>();
  useEffect(() => {
    ref.current = value;
  });
  return ref.current;
};

const _utilityFunctions = {
  isDualScreen,
  usePrevious
};

type utilityFunctions = typeof _utilityFunctions;

const utility: utilityFunctions = _utilityFunctions;

export default utility;
