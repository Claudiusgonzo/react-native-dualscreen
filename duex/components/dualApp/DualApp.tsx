import React, { Fragment, useEffect } from 'react';
import { Provider } from 'react-redux';

import { store } from '../../appStore';
import { IScreenComponent } from '../../utilities/interfaces';
import DualHub from '../dualHub/DualHub';
import singleScreen from '../../singleScreen/singleScreenStore/singleScreen.methods';
import dualScreen from '../../dualScreen/dualScreenStore/dualScreen.methods';

export interface IDualAppProps {
  /**
   * Default element for single screen mode
   */
  singleScreen: IScreenComponent;

  /**
   * Default element for dual screen mode
   */
  dualScreen: IScreenComponent;

  /**
   * AppContainer if using React Navigation
   */
  navigationContainer?: JSX.Element;

  /**
   * Items to render in app drawer. If empty, drawer will not be included.
   */
  drawerItems?: JSX.Element[];
}

const DualApp = (props: IDualAppProps) => {

  const screenState = store.getState().KeyReducers.keys;
  //TODO: possibly look into utility state for things like isMount to keep data between renders
  const hasSingle: boolean = screenState.some(val => val.key === (`SINGLE_${props.singleScreen.key}`));
  const hasDual: boolean = screenState.some(val => val.key === (`DUAL_${props.singleScreen.key}`));

  useEffect(() => {
    if (!hasSingle && !hasDual) {
      singleScreen.Add(props.singleScreen.key, props.singleScreen.duexElement, props?.singleScreen?.header!);
      dualScreen.Add(props.dualScreen.key, props.dualScreen.duexElement, props?.dualScreen?.header!);
    }
  }, []);

  return (
    <Fragment key={'App Component'}>
      <Provider store={store}>
        {props.navigationContainer !== undefined ? (
          props.navigationContainer
        ) : (
            <Fragment>
              <DualHub />
            </Fragment>
          )}
      </Provider>
    </Fragment>
  );
};

export default DualApp;
