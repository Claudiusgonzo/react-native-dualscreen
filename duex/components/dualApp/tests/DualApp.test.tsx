import React, { Fragment } from 'react';
import DualApp from '../DualApp';
import { IScreenComponent } from '../../../utilities/interfaces';
import { render, toJSON } from '@testing-library/react-native';

describe('DualApp Tests', () => {
    it('should render singleScreen', () => {
        // Arrange
        jest.useFakeTimers();
        const singleScreenDefault: IScreenComponent = {
            key: 'singleDefault',
            duexElement: <Fragment />,
            header: {
                title: 'singleScreenDefaultTitle'
            }
        };

        const dualScreenDefault: IScreenComponent = {
            key: 'dualDefault',
            duexElement: <Fragment />,
            header: {
                title: 'dualScreenDefaultTitle'
            }
        };

        // Act
        const { container } = render(
            <DualApp
                singleScreen={singleScreenDefault}
                dualScreen={dualScreenDefault}
            />
        );

        // Assert
        expect(toJSON(container)).toMatchSnapshot();
    });
});
