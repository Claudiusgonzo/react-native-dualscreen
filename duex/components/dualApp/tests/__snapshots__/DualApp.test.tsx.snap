// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DualApp Tests should render singleScreen 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "black",
            "bottom": 0,
            "flex": 1,
            "height": 720,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 540,
          },
          Object {
            "left": 0,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "height": "10%",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                },
                Object {
                  "backgroundColor": "#212121",
                  "paddingHorizontal": 25,
                },
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#F2F2F2",
                  "fontSize": 20,
                  "fontWeight": "500",
                  "marginLeft": 6,
                }
              }
            >
              singleScreenDefaultTitle
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "height": "85%",
          }
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "black",
            "bottom": 0,
            "flex": 1,
            "height": 720,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 540,
          },
          Object {
            "left": 570,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "height": "10%",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                },
                Object {
                  "backgroundColor": "#212121",
                  "paddingHorizontal": 25,
                },
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#F2F2F2",
                  "fontSize": 20,
                  "fontWeight": "500",
                  "marginLeft": 6,
                }
              }
            >
              dualScreenDefaultTitle
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "height": "85%",
          }
        }
      />
    </View>
  </View>
</View>
`;
